### PR TITLE
Better loss tracking

### PR DESCRIPTION
**Pull request by kaminow**
Change the stats that are tracked logged during training to include the compound and label associated with loss values, rather than just the loss values themselves. This will make it easier to figure out what's going on in training and isolate any specific molecules that might be difficult for the model.

Also some other minor changes that coincided with this, mainly a rearrangement of the code in `train.py` to be smarter about `wandb` initialization and letting it play better with CLI args.

### CODE REVIEW

File: asapdiscovery-ml/asapdiscovery/ml/scripts/plot_data_efficiency.py [hunk 1]
```diff
@@ -147,6 +148,104 @@ def losses_to_df(losses):
...
+        df = df.loc[df["epoch"] == best_idx, :]
+
+    # Convert losses if desired
+    if conv_function:
+        df.loc[:, "loss"] = df["loss"].pow(0.5).apply(conv_function)
+
+    return df
+
+
+def load_all_losses_dict(in_df, rel_dir=None, conv_function=None, take_best=True):
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/plot_data_efficiency.py:207]: Perhaps consider the name collision between this and `def load_losses_dict`, these two were close enough to be a bit confusing.

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 2]
```diff
@@ -146,20 +145,60 @@ def make_wandb_table(ds_split):
 def wandb_init(
     project_name,
     run_name,
-    exp_configure,
-    ds_splits,
-    ds_split_labels=["train", "val", "test"],
+    sweep,
+    cont,
+    model_o=None,
+    exp_configure=None,
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:151]: Docstring, what are these params?

File: asapdiscovery-ml/asapdiscovery/ml/scripts/train.py [hunk 3]
```diff
@@ -651,28 +721,19 @@ def main():
...
         exp_configure,
+        model_dir,
     ) = init(args)
 
     # Load model weights as necessary
     if args.cont:
         start_epoch, wts_fn = find_most_recent(args.model_o)
 
         # Load error dicts
-        if os.path.isfile(f"{args.model_o}/train_err.pkl"):
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/scripts/train.py:661]: Good work getting rid of the pickles. Json much better.

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 4]
```diff
@@ -1475,12 +1485,18 @@ def train(
     if use_wandb:
         import wandb
 
-    if train_loss is None:
-        train_loss = []
-    if val_loss is None:
-        val_loss = []
-    if test_loss is None:
-        test_loss = []
+    # Dicts of compounds (one for each split) storing true label and
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:1488]: Remove commented stuff if not needed.

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 5]
```diff
@@ -1533,6 +1552,19 @@ def train(
             pred = model(pose).reshape(target.shape)
             loss = loss_fn(pred, target, in_range, uncertainty)
 
+            # Update loss_dict
+            try:
+                loss_dict["train"][compound_id]["preds"].append(pred.item())
+                loss_dict["train"][compound_id]["losses"].append(loss.item())
+            except KeyError:
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:1559]: Comment as to what exactly the logic is here?

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 6]
```diff
@@ -1585,8 +1616,21 @@ def train(
                 # Make prediction and calculate loss
                 pred = model(pose).reshape(target.shape)
                 loss = loss_fn(pred, target, in_range, uncertainty)
+
+                # Update loss_dict
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:1620]: Repetition, use some kind of function or abstraction.

File: asapdiscovery-ml/asapdiscovery/ml/utils.py [hunk 7]
```diff
@@ -1610,8 +1654,21 @@ def train(
...
```
Comment (hmacdope) [inline asapdiscovery-ml/asapdiscovery/ml/utils.py:1658]: As above.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking good, just a few queries and suggestions.
